src/Features/CSharp/Portable/CodeMapper/CSharpMapCodeService.NodeToMap.cs

```diff
@@ -0,0 +1,423 @@
+        {
+            var firstDeclarator = fieldDeclaration.Declaration.Variables.FirstOrDefault();
+            if (firstDeclarator is not null)
+            {
+                return firstDeclarator.Identifier.ToString();
+            }
+
+            // Fallback to returning the full string of the node as identifier.
+            return fieldDeclaration.ToString();
+        }
+
+        private static string BuildLocalFunctionIdentifier(LocalFunctionStatementSyntax localFunction)
+        {
+            var identifier = localFunction.Identifier.ToString();
+
+            // when identifier is empty, that means this
+            // local function declaration represents a Constructor
+            if (string.IsNullOrWhiteSpace(identifier))
+            {
+                return BuildConstructorIdentifier(localFunction);
+            }
+
+            return identifier;
+        }
+
+        /// <summary>
+        /// Returns a composed version of the constructor's identifier.
+        /// The string is composed by the original identifier of the constructor
+        /// and the types of its parameters.
+        /// Example:
+        /// MyClass(string, string, int)
+        /// </summary>
+        /// <param name="ctor">The constructor declaration syntax</param>
+        /// <returns>An identifier string composed by the constructor's identifier and argument types.</returns>
+        private static string BuildConstructorIdentifier(SyntaxNode ctor)
+        {
+            ParameterListSyntax parameters;
+            var identifier = string.Empty;
+            if (ctor is ConstructorDeclarationSyntax cds)
+            {
+                identifier = cds.Identifier.ToString();
+                parameters = cds.ParameterList;
+            }
+            else if (ctor is LocalFunctionStatementSyntax lfs)
+            {
+                identifier = lfs.ReturnType.ToString();
+                parameters = lfs.ParameterList;
+            }
+            else
+            {
+                return ctor.ToString();
+            }
+
+            var parameterTypes = parameters.Parameters
+                .Select(p => p.Type?.ToString())
+                .Where(type => type is not null);
+            if (parameterTypes.Any())
+            {
+                return $"{identifier}({string.Join(", ", parameterTypes)})";
+            }
+            else
+            {
+                return $"{identifier}()";
+            }
+        }
+
+        /// <summary>
+        /// Extracts source nodes from a code snippet.
+        /// </summary>
+        public static async Task<ImmutableArray<NodeToMap>> ExtractSourceNodesAsync(string code, CSharpParseOptions options, CancellationToken cancellationToken)
+        {
+            var syntaxTree = CSharpSyntaxTree.ParseText(SourceText.From(code), options, cancellationToken: cancellationToken);
+            var rootNode = await syntaxTree.GetRootAsync(cancellationToken).ConfigureAwait(false);
```

Comment:
this doesn't need to be async.  you can just call GetRoot.  since you just parsed teh tree, it's o(1).